plugins/filter_aws/aws.c

```diff
@@ -698,23 +700,89 @@ static int get_ec2_metadata(struct flb_filter_aws *ctx)
             flb_plg_error(ctx->ins, "Failed to get Hostname");
             return -1;
         }
-        ctx->new_keys++;
     }
 
+    return 0;
+}
+
+static int get_ec2_metadata_tags(struct flb_filter_aws *ctx)
+{
+    int ret;
+
+    ctx->group_tag.
+        last_fetch_attempt = time(NULL);
+
     if (ctx->tags_enabled && !ctx->tags_fetched) {
         ret = get_ec2_tags(ctx);
         if (ret < 0) {
             flb_plg_error(ctx->ins, "Failed to get instance EC2 Tags");
             return ret;
         }
-        for (i = 0; i < ctx->tags_count; i++) {
-            if (ctx->tag_is_enabled[i] == FLB_TRUE) {
-                ctx->new_keys++;
+    }
+
+    return 0;
+}
+
+static int ec2_metadata_group_should_fetch(struct flb_filter_aws_metadata_group *group)
+{
+    time_t now, required_interval, interval;
+
+    required_interval = group->retry_required_interval;
+    if (required_interval == 0) {
+        return FLB_TRUE;
+    }
+
+    now = time(NULL);
+
+    interval = now - group->last_fetch_attempt;
+
+    if (interval < required_interval) {
+        return FLB_FALSE;
+    }
+    return FLB_TRUE;
+}
+
+/*
+ * Fetches all metadata values, including tags, from IMDS.
+ * Function handles retries as configured for each metadata group.
+ *
+ * Returns 0 on success, negative values on failures.
+ * Returns FLB_FILTER_AWS_CONFIGURATION_ERROR in case of configuration error.
+ */
+static int get_ec2_metadata(struct flb_filter_aws *ctx)
+{
+    int ret;
+    int no_failures = FLB_TRUE;
+    int no_fetches_skipped = FLB_TRUE;
```

Comment:
do we really need two bools here? Could it just be one "is_metadata_fetched" or shorter "has_metadata"

Reply:
Yeah, you are right. We can have just one bool.

I modified the code and there is just one boolean variable now: `metadata_fetched`.